### ISSUE TITLE
BadValue $in needs an array is misleading

### ISSUE THREAD

**Issue by orestes** [doctrine/mongodb-odm#1122]
I just updated my dependencies and found a BC break

composer.json
```json
    "doctrine/mongodb-odm": "1.0.*@dev",
    "doctrine/mongodb-odm-bundle": "3.0.*@dev",
```
composer.lock
```json
{
"name": "doctrine/mongodb-odm",
            "version": "dev-master",
            "source": {
                "type": "git",
                "url": "https://github.com/doctrine/mongodb-odm.git",
                "reference": "d305acad9307a9f7419ec8a6f904de2fae38f770"
              }
}
```
I'm using mongodb-odm in a Symfony2 app.

I just ran into a problem that I think was caused by mongodb-odm. We ere using an associative array for a term search that looks like this

```php
        $users = $repository->findBy([
            'field' => ['$in' => $values]
        ]);
```

Previously to my update, the $values array was handled fine even if the keys were non-consecutive numbers, but that broke with the update.

I fixed by using only the values and discarding the keys since they're not needed.

```php
        $values = array_values($original_values);
        $users = $repository->findBy([
            'field' => ['$in' => $values]
        ]);
```

I'm thinking this might have more to do with our update from MongoDB 2.6 to 3.03. Maybe I was wrong thinking this was a mongodb-odm issue.

**Comment by jmikola** [comment]
**tl;dr**: use `array_values()` in your free-form query criteria and update documents when MongoDB actually expects a BSON array.

This was primarily addressed in doctrine/mongodb#174, where we fixed the query builder to sanitize user input through `array_values()` (in order to ensure the value would encode as a BSON array). That fix was released in 1.1.6; however, it only pertains to the query builder. Doctrine MongoDB and the ODM, which depends upon it, do not make any attempt to sanitize free-form query criteria (like you have here). ODM does do some translation on the field names when the DB and PHP mappings differ, but that's just on the keys of query/update documents.

AFAIK, There was *one* case in ODM was internally crafting an `$in` query without running it through `array_values()`, but that was addressed in #743.

Per my comment in https://github.com/doctrine/mongodb/pull/174#issuecomment-40825006, this was also reported against the PHP driver in [PHP-1051](https://jira.mongodb.org/browse/PHP-1051); however, this strict validation appeared in MongoDB 2.6, so I'm not sure why you only noticed it upon upgrading from 2.6 to 3.0.3. Hopefully the solution makes sense, though. It's an unfortunate side effect of PHP using arrays for both lists and dictionaries, and legacy design decisions in the 1.x driver :)

**Comment by orestes** [comment]
Thanks for the superb explanation @jmikola 
I've been checking some facts and we actually updated from an even older version installed from the Debian Wheezy repos which were really outdated.

Sorry for the noise and thank you again for your explanation :)

**Comment by jmikola** [comment]
![](http://www.reactiongifs.com/r/tu.gif)

No problem! Glad to hear it didn't magically show up between 2.6 and 3.0.3 -- now I can sleep at night :)
